plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -1710,7 +1858,35 @@ public void rollback()
     @Override
     public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(ConnectorSession session, ConnectorTableHandle handle, Constraint constraint)
     {
-        IcebergTableHandle table = (IcebergTableHandle) handle;
+        if (handle instanceof IcebergTableHandle icebergTableHandle) {
+            return applyDataFilter(session, icebergTableHandle, constraint);
+        }
+        else if (handle instanceof IcebergSystemTableHandle icebergSystemTableHandle) {
+            return applySystemFilter(icebergSystemTableHandle, constraint);
+        }
+        else {
+            throw new IllegalArgumentException("Unknown connector table handle type '" + handle.getClass().getName() + "'");
+        }
+    }
+
+    private Optional<ConstraintApplicationResult<ConnectorTableHandle>> applySystemFilter(IcebergSystemTableHandle table, Constraint constraint)
+    {
+        TupleDomain<ColumnHandle> oldDomain = table.getConstraint();
+        TupleDomain<ColumnHandle> newDomain = oldDomain.intersect(constraint.getSummary());
+        if (oldDomain.equals(newDomain) && constraint.predicate().isEmpty()) {
+            return Optional.empty();
+        }
+
+        if (oldDomain.equals(newDomain)) {
```

Comment:
In `io.trino.connector.system.SystemTablesMetadata#applyFilter` there is made the call 

https://github.com/trinodb/trino/blob/3e06ac2c94da4364fc386c7b0b3f82460e406a5e/core/trino-main/src/main/java/io/trino/connector/system/SystemTablesMetadata.java#L159

to verify whether the table has disappeared in the meantime.

Should we do it here as well?